source/opt/dominator_tree.h

```diff
@@ -0,0 +1,180 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_DOMINATOR_ANALYSIS_TREE_H_
+#define LIBSPIRV_OPT_DOMINATOR_ANALYSIS_TREE_H_
+
+#include <cstdint>
+#include <map>
+#include <utility>
+#include <vector>
+
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+// This helper struct forms the nodes in the tree, with each node containing its
+// children. It also contains two values, for the pre and post indexes in the
+// tree which are used to compare two nodes.
+struct DominatorTreeNode {
+  explicit DominatorTreeNode(ir::BasicBlock* bb)
+      : bb_(bb),
+        parent_(nullptr),
+        childrens_({}),
+        dfs_num_pre_(-1),
+        dfs_num_post_(-1) {}
+
+  inline uint32_t id() const { return bb_->id(); }
+
+  ir::BasicBlock* bb_;
+  DominatorTreeNode* parent_;
+  std::vector<DominatorTreeNode*> childrens_;
+
+  // These indexes are used to compare two given nodes. A node is a child or
+  // grandchild of another node if its preorder index is greater than the
+  // first nodes preorder index AND if its postorder index is less than the
+  // first nodes postorder index.
+  int dfs_num_pre_;
+  int dfs_num_post_;
+};
+
+// A class representing a tree of BasicBlocks in a given function, where each
+// node is dominated by its parent.
+class DominatorTree {
+ public:
+  // Map OpLabel ids to dominator tree nodes
+  using DominatorTreeNodeMap = std::map<uint32_t, DominatorTreeNode>;
+  using iterator = DominatorTreeNodeMap::iterator;
+  using const_iterator = DominatorTreeNodeMap::const_iterator;
+
+  // List of DominatorTreeNode to define the list of roots
+  using DominatorTreeNodeList = std::vector<DominatorTreeNode*>;
+  using roots_iterator = DominatorTreeNodeList::iterator;
+  using roots_const_iterator = DominatorTreeNodeList::const_iterator;
+
+  DominatorTree() : postdominator_(false) {}
+  explicit DominatorTree(bool post) : postdominator_(post) {}
+
+  iterator begin() { return nodes_.begin(); }
+  iterator end() { return nodes_.end(); }
+  const_iterator begin() const { return cbegin(); }
+  const_iterator end() const { return cend(); }
+  const_iterator cbegin() const { return nodes_.begin(); }
+  const_iterator cend() const { return nodes_.end(); }
+
+  roots_iterator roots_begin() { return roots_.begin(); }
+  roots_iterator roots_end() { return roots_.end(); }
+  roots_const_iterator roots_begin() const { return roots_cbegin(); }
+  roots_const_iterator roots_end() const { return roots_cend(); }
+  roots_const_iterator roots_cbegin() const { return roots_.begin(); }
+  roots_const_iterator roots_cend() const { return roots_.end(); }
+
+  // Get the unique root of the tree.
```

Comment:
It seems like some of this would be simpler using both a pseudo entry (which is used) and a pseudo exit block. Then there would be a single root for both cases and that root would never be a true basic block. Opinions?

Reply:
That's actually how we implemented it at first but we regard the pseudo entry node an implementation detail of the algorithm and as such avoided exposing it as much as possible. It also means that every tree in a module would share the same root because the pseudo entry/exit blocks are retrieved from the cfg on a module scope which could lead to some confusion.